File: java/creational/PrototypePattern.java
package creational;

/**
 * Prototype Pattern
 * Creates new objects by cloning existing ones.
 */
public class PrototypePattern {
    
    // Prototype interface
    interface Prototype extends Cloneable {
        Prototype clone();
    }
    
    // Concrete Prototype
    static class Shape implements Prototype {
        private String type;
        private String color;
        private int x;
        private int y;
        
        public Shape(String type, String color, int x, int y) {
            this.type = type;
            this.color = color;
            this.x = x;
            this.y = y;
        }
        
        // Copy constructor
        public Shape(Shape source) {
            this.type = source.type;
            this.color = source.color;
            this.x = source.x;
            this.y = source.y;
        }
        
        @Override
        public Shape clone() {
            return new Shape(this);
        }
        
        public void setPosition(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        @Override
        public String toString() {
            return "Shape{type='" + type + "', color='" + color + 
                   "', position=(" + x + "," + y + ")}";
        }
    }
    
    // Prototype Registry
    static class ShapeRegistry {
        private java.util.Map<String, Shape> shapes = new java.util.HashMap<>();
        
        public void addShape(String key, Shape shape) {
            shapes.put(key, shape);
        }
        
        public Shape getShape(String key) {
            Shape prototype = shapes.get(key);
            return prototype != null ? prototype.clone() : null;
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Prototype Pattern Demo ===\n");
        
        // Create original shapes
        Shape circle = new Shape("Circle", "Red", 10, 20);
        Shape rectangle = new Shape("Rectangle", "Blue", 30, 40);
        
        System.out.println("Original shapes:");
        System.out.println(circle);
        System.out.println(rectangle);
        
        // Clone shapes
        Shape clonedCircle = circle.clone();
        clonedCircle.setPosition(50, 60);
        
        Shape clonedRectangle = rectangle.clone();
        clonedRectangle.setPosition(70, 80);
        
        System.out.println("\nCloned shapes (with modified positions):");
        System.out.println(clonedCircle);
        System.out.println(clonedRectangle);
        
        System.out.println("\nOriginal shapes (unchanged):");
        System.out.println(circle);
        System.out.println(rectangle);
        
        // Using registry
        System.out.println("\n--- Using Prototype Registry ---");
        ShapeRegistry registry = new ShapeRegistry();
        registry.addShape("red-circle", new Shape("Circle", "Red", 0, 0));
        registry.addShape("blue-rectangle", new Shape("Rectangle", "Blue", 0, 0));
        
        Shape shape1 = registry.getShape("red-circle");
        shape1.setPosition(100, 100);
        
        Shape shape2 = registry.getShape("red-circle");
        shape2.setPosition(200, 200);
        
        System.out.println("Shape 1: " + shape1);
        System.out.println("Shape 2: " + shape2);
    }
}

File: python/creational/prototype_pattern.py
"""
Prototype Pattern
Purpose: Creates new objects by cloning existing ones
Use Case: Object creation is expensive, need to avoid subclassing
"""

import copy
from abc import ABC, abstractmethod
from typing import Dict


class Prototype(ABC):
    @abstractmethod
    def clone(self):
        pass


class Document(Prototype):
    def __init__(self, title: str, content: str, metadata: dict):
        self.title = title
        self.content = content
        self.metadata = metadata
    
    def clone(self):
        # Deep copy for mutable attributes
        return copy.deepcopy(self)
    
    def __str__(self):
        return f"Document(title='{self.title}', metadata={self.metadata})"


class Shape(Prototype):
    def __init__(self, x: int, y: int, color: str):
        self.x = x
        self.y = y
        self.color = color
    
    def clone(self):
        return copy.copy(self)


class Circle(Shape):
    def __init__(self, x: int, y: int, color: str, radius: int):
        super().__init__(x, y, color)
        self.radius = radius
    
    def __str__(self):
        return f"Circle(x={self.x}, y={self.y}, color='{self.color}', radius={self.radius})"


class Rectangle(Shape):
    def __init__(self, x: int, y: int, color: str, width: int, height: int):
        super().__init__(x, y, color)
        self.width = width
        self.height = height
    
    def __str__(self):
        return f"Rectangle(x={self.x}, y={self.y}, color='{self.color}', width={self.width}, height={self.height})"


class PrototypeRegistry:
    """Registry of prototypes for easy cloning"""
    def __init__(self):
        self._prototypes: Dict[str, Prototype] = {}
    
    def register(self, key: str, prototype: Prototype):
        self._prototypes[key] = prototype
    
    def unregister(self, key: str):
        del self._prototypes[key]
    
    def clone(self, key: str) -> Prototype:
        return self._prototypes[key].clone()


if __name__ == "__main__":
    print("=== Prototype Pattern Demo ===\n")
    
    # Document cloning
    doc1 = Document("Report", "Content here", {"author": "Alice", "version": 1})
    doc2 = doc1.clone()
    doc2.title = "Report Copy"
    doc2.metadata["version"] = 2
    
    print(f"Original: {doc1}")
    print(f"Clone: {doc2}")
    
    # Shape cloning
    circle1 = Circle(10, 20, "red", 5)
    circle2 = circle1.clone()
    circle2.color = "blue"
    
    print(f"\nOriginal: {circle1}")
    print(f"Clone: {circle2}")
    
    # Registry usage
    registry = PrototypeRegistry()
    registry.register("basic_circle", Circle(0, 0, "black", 10))
    registry.register("basic_rect", Rectangle(0, 0, "white", 50, 30))
    
    new_circle = registry.clone("basic_circle")
    print(f"\nCloned from registry: {new_circle}")
